subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultFileCollectionSnapshot.java

```diff
@@ -79,12 +100,23 @@ public boolean isEmpty() {
 
     @Override
     public Iterator<TaskStateChange> iterateContentChangesSince(FileCollectionSnapshot oldSnapshot, String fileType, boolean includeAdded) {
+        if (includeAdded && hashCode != null && getHash().equals(oldSnapshot.getHash())) {
+            return Iterators.emptyIterator();
+        }
         return compareStrategy.iterateContentChangesSince(snapshots, oldSnapshot.getSnapshots(), fileType, pathIsAbsolute, includeAdded);
     }
 
     @Override
+    public HashCode getHash() {
+        if (hashCode == null) {
+            hashCode = hashCodeSupplier.get();
+        }
+        return hashCode;
```

Comment:
What is the use of the memoizing `Supplier` if we also do a `null`-check here? If you want to use a `Supplier`, you could wrap the `hashCode` in a constant one if it's provided in the constructor, or use a calculating one if it's not. We could use our own `Factory`, too.

But I would simply write the code out here. I don't think it's worth creating all these objects just to hide that small amount of code.

Reply:
I removed the memoizing supplier and inlined the code.